File: backend/gatorsched_api/schemas/employee/requests/employee_requests.py
from typing import Annotated, Literal

from pydantic import BaseModel, Field

from gatorsched_api.models.callout_request import CallOutStatus
from gatorsched_api.models.types import EmployeeRequestStatus, ManagerRequestStatus

RequestType = Literal["swap", "callout", "pickup"]


class RequestPerson(BaseModel):
    id: str
    name: str
    avatarUrl: str | None = None
    color: str


class RequestShiftSummary(BaseModel):
    dayLabel: str
    timeRange: str


class EmployeeRequestCardBase(BaseModel):
    id: str
    employeeStatus: EmployeeRequestStatus
    managerStatus: ManagerRequestStatus


class EmployeeSwapRequestCard(EmployeeRequestCardBase):
    type: Literal["swap"] = "swap"
    requester: RequestPerson
    requesterShift: RequestShiftSummary
    coverEmployee: RequestPerson
    coverShift: RequestShiftSummary

class EmployeeCalloutRequestCard(EmployeeRequestCardBase):
    type: Literal["callout"] = "callout"
    employee: RequestPerson
    shift: RequestShiftSummary
    reason: str | None = None
    status: CallOutStatus

# Placeholder alias until additional employee request card types are introduced
EmployeeRequestCard = Annotated[
    EmployeeSwapRequestCard | EmployeeCalloutRequestCard,
    Field(discriminator="type"),
]


class EmployeeRequestsResponse(BaseModel):
    incoming: list[EmployeeRequestCard]
    outgoing: list[EmployeeRequestCard]

File: frontend/src/services/requests/employeeRequests.ts
import { API_BASE } from '../apiConfig';

import {
  EmployeeRequestStatus,
  ManagerRequestStatus,
  RequestPerson,
  RequestShiftSummary,
} from './types';

export type EmployeeRequestCardBase = {
  id: string;
  employeeStatus: EmployeeRequestStatus;
  managerStatus: ManagerRequestStatus;
};

export type EmployeeSwapRequestCard = EmployeeRequestCardBase & {
  type: 'swap';
  requester: RequestPerson;
  requesterShift: RequestShiftSummary;
  coverEmployee: RequestPerson;
  coverShift: RequestShiftSummary;
};

export type EmployeeCalloutRequestCard = EmployeeRequestCardBase & {
  type: 'callout';
  employee: RequestPerson;
  shift: RequestShiftSummary;
  reason: string | null;
};

export type EmployeeRequestCard = EmployeeSwapRequestCard | EmployeeCalloutRequestCard;

export type EmployeeRequestsResponse = {
  incoming: EmployeeRequestCard[];
  outgoing: EmployeeRequestCard[];
};

export async function getEmployeeRequests(viewerId: string): Promise<EmployeeRequestsResponse> {
  const res = await fetch(
    `${API_BASE}/employee/requests?viewer_id=${encodeURIComponent(viewerId)}`,
  );

  if (!res.ok) {
    throw new Error('Failed to retrieve Employee Requests');
  }

  return await res.json();
}
